### PR TITLE
Added recursive shallow-clone support for git-submodules (#2123)

### PR DESCRIPTION
**Pull request by GaneshSPatil**
When shallow clone option is provided, all the submodules will be shallow-cloned recursively (if a submodule contains other submodules, those will also be shallow-cloned).

### CODE REVIEW

**Comment by GaneshSPatil** [comment]
Submodules 'shallow-clone' feature is supported only if the git version >= **2.9.0**
if git version < 2.9.0 submodules are not shallow cloned.

File: common/src/com/thoughtworks/go/domain/materials/git/GitCommand.java [hunk 1]
```diff
@@ -69,14 +71,28 @@ public int clone(ProcessOutputStreamConsumer outputStreamConsumer, String url) {
...
         }
         gitClone.withArg(new UrlArgument(url)).withArg(workingDir.getAbsolutePath());
+        return gitClone;
+    }
 
-        return run(gitClone, outputStreamConsumer);
+    protected boolean isSubModuleShallowCloneSupported() {
+        String version = GitCommand.version(new HashMap<String, String>());
+        //--shallow-submodules flag is supported from git version 2.9.0
+        return GIT_SHALLOW_SUBMODULE_CLONE_SUPPORTED_PATTERN.matcher(version).find();
```
Comment (ketan) [inline common/src/com/thoughtworks/go/domain/materials/git/GitCommand.java]: See `parseGitVersion`

File: domain/src/com/thoughtworks/go/domain/materials/git/GitCommand.java [hunk 2]
```diff
@@ -69,14 +72,45 @@ public int clone(ProcessOutputStreamConsumer outputStreamConsumer, String url) {
...
+        CommandLine gitClone = getGitCloneCommand(outputStreamConsumer, url, depth);
+        return run(gitClone, outputStreamConsumer);
+    }
+
+    protected CommandLine getGitCloneCommand(ProcessOutputStreamConsumer outputStreamConsumer, String url, Integer depth) {
         CommandLine gitClone = cloneCommand();
 
         if(depth < Integer.MAX_VALUE) {
             gitClone.withArg(String.format("--depth=%s", depth));
+            if(isSubModuleShallowCloneSupported()){
```
Comment (varshavaradarajan) [inline domain/src/com/thoughtworks/go/domain/materials/git/GitCommand.java]: Indent code?

File: domain/src/com/thoughtworks/go/domain/materials/git/GitCommand.java [hunk 3]
```diff
@@ -69,14 +72,45 @@ public int clone(ProcessOutputStreamConsumer outputStreamConsumer, String url) {
...
+    }
 
-        return run(gitClone, outputStreamConsumer);
+    protected boolean isSubModuleShallowCloneSupported() {
+        String version = GitCommand.version(new HashMap<String, String>());
+        //--shallow-submodules flag is supported from git version 2.9.0
+        return isVersionEqualToOrHigherThan(version, 2.9f);
+    }
+
+    public static boolean isVersionEqualToOrHigherThan(String hgout, float version) {
```
Comment (varshavaradarajan) [inline domain/src/com/thoughtworks/go/domain/materials/git/GitCommand.java]: Change the argument name to something that indicates that its the output of the command git version.

File: domain/src/com/thoughtworks/go/domain/materials/git/GitCommand.java [hunk 4]
```diff
@@ -69,14 +72,45 @@ public int clone(ProcessOutputStreamConsumer outputStreamConsumer, String url) {
...
+        String hgVersion = parseGitVersion(hgout);
```
Comment (varshavaradarajan) [inline domain/src/com/thoughtworks/go/domain/materials/git/GitCommand.java]: Same as above.

File: common/test/unit/com/thoughtworks/go/domain/materials/git/GitCommandTest.java [hunk 5]
```diff
@@ -352,6 +403,41 @@ public void shouldReturnFalseIfTheGivenBranchDoesNotContainTheRevision() {
     }
 
     @Test
+    public void shouldReturnTrueIfVersionHigherThan1dot6Linux() throws Exception {
+        assertThat(git.isVersionEqualToOrHigherThan(GIT_VERSION_1_6_0_2, 1.6f), Is.is(true));
+    }
+
+    @Test
+    public void shouldReturnTrueIfVersionLowerThanExpectedLinux() throws Exception {
```
Comment (varshavaradarajan) [inline common/test/unit/com/thoughtworks/go/domain/materials/git/GitCommandTest.java]: You are asserting on false while saying shouldReturnTrue. Please change the name of the test to indicate what the test is checking for as far as possible.

File: common/test/unit/com/thoughtworks/go/domain/materials/git/GitCommandTest.java [hunk 6]
```diff
@@ -352,6 +403,41 @@ public void shouldReturnFalseIfTheGivenBranchDoesNotContainTheRevision() {
...
+    public void shouldReturnTrueIfVersionLowerThan2dot9Linux() throws Exception {
+        assertThat(git.isVersionEqualToOrHigherThan(GIT_VERSION_1_5_4_3, 2.9f), Is.is(false));
+    }
+
+    @Test
+    public void shouldReturnTrueIfVersionHigherThan2dot9Windows() throws Exception {
+        assertThat(git.isVersionEqualToOrHigherThan(GIT_VERSION_2_9_0_2_ON_WINDOWS, 2.9f), Is.is(true));
+    }
+
+    @Test(expected = Exception.class)
```
Comment (varshavaradarajan) [inline common/test/unit/com/thoughtworks/go/domain/materials/git/GitCommandTest.java]: Is there a specific exception that you'd need to be checking for?

**Comment by varshavaradarajan** [comment]
Hey, can you please change the copyright of these files to 2017?

**Comment by GaneshSPatil** [comment]
Version matrix —

Git version | flag          | description
------------|---------------|--------------------------------------------------------------------------------------
1.7.1       | `--depth`     | shallow clone
1.7.1       | `--recursive` | clone all submodules recursively
2.8.0       | `--jobs=4`    | number of fetch jobs on `git fetch`. Also `submodule.fetchJobs` config option (safer)

**Comment by GaneshSPatil** [comment]
need to discuss with @ketan about it,
keeping it open for now.

**Comment by varshavaradarajan** [comment]
To be decided whether to reopen after discussion.

**Comment by ketan** [comment]
@GaneshSPatil — can you recall why this was closed? I can see from the release notes that the minimum git version requirement is now 1.9, however we don't seem to leverage most of the CLI support with that version of git.

**Comment by GaneshSPatil** [comment]
@ketan -- from what I remember, one of the reasons was, we wanted support managing keys hence we said it would be better to do it as part of the plugin rather than core.

**Pull request closed without merging**